investigations/LOKI/loki_performance.md

```diff
@@ -0,0 +1,122 @@
+
+- 64GB RAM
+- 7th Gen Intel Core i7 Quad Core (8 threads) 8MB SmartCache
+
+Subsequently, further optimizations to the Mantid live reduction code was performed by Dan Nixon. The detail of which is outlined below:
+
+*Instead of received events (from the Kafka stream) being immediately inserted into an EventWorkspace they are instead stored in a buffer held by the decoder.
+When this buffer is full (where full is a number of events set by the user) it is sorted to place events from the same detector in the same period adjacent to one another. This locality is then used to allow those events to be inserted into the EventWorkspace in parallel (as division of the work across threads can guarantee no EventList is accessed on multiple threads).*
+
+*Instead of storing a hash map of detector to workspace index (which is slow to index), a vector is created where the indices represent detector numbers and the values are the associated workspace index (which is much faster to index but would typically use more memory).*
+
+These optimizations pushed the achievable kafka message consumption rate to >20Hz which satisfies, and exceeds, the 14Hz target.
+
+#### Conclusion
+Even with random data (noise), which represents the worst case scenario for event collection, Mantid is able to cope with an event rate of 10<sup>7</sup>Hz. **N.B** This is only for writing events to the Workspace and performing a simple rebin to histogram for the output workspace.
+
+### MCStas/Geant4 Simulated Data Test 
+
+Realistic LOKI data was generated by Judith Houston (ESS LoKI instrument scientist) using McStas and Geant4. This simulated data was ingested into Mantid to produce a Nexus file which could be used in the `NeXus-Streamer` for our test purposes. The dataset consists of neutrons scattered on the rear LOKI panel ignoring the front 2 banks (as pictured below).
+
+![LOKI Geant4 data](loki_geant4.png)
+
+The data file contained ~1M events which represented a single pulse. In order to generate a longer run, data was duplicated over 840 pulses (60 second run at 14Hz). This data was then streamed into Kafka and used to test consumption rates and data reduction rates.
+
+#### Writing To Workspace
+
+Data consumption rates averaged ~24Hz (messages) at an event rate of ~1.4e<sup>7</sup>Hz. This utilized a 30s timeout for `MonitorLiveData` and a basic `Rebin` operation for displaying data on the instrument. The event buffer stored 2.5e<sup>7</sup> events before writing to the workspace.
+
+##### Scaling
+There was previous work done in investigating how well the live streamer in Mantid scales and was presented in this document [LiveReductionInvestigation](https://github.com/DMSC-Instrument-Data/documents/blob/master/investigations/Live%20Reduction/LiveReductionInvestigation.md). Scaling for LOKI ingesting the simulated data showed the following trend:
+
+`#` of threads|Message Consumption Rate
+---|---
```

Comment:
Yes Dan is getting these figures

Reply:
@DanNixon Should these numbers be divided by the number of banks?

Reply:
Maybe it is easier to present it that way. I left it as messages since that's what the header stated and this measurement is less instrument specific.

Yes, dividing by number of banks (9) will give the maximum pulse rate which is probably a clearer indication of whether or not LOKI can be streamed.

Reply:
Ok I will do that as I am making changes to the document. Thanks Dan